converted the slave to master access control to github (#3317)
jenkins-infra/jenkins.io: content/doc/book/using/slave-master-access-control.adoc

---
layout: section
title: Agent to master access control
---

> [[SlaveToMasterAccessControl-Whatisthis]]

Reviewer: I'd suggest to simplify the section names, removing the common prefix
```suggestion
[[Whatisthis]]
```

Reviewer 2: The odd section names may be a result of the wiki exporter working to help us retain working hyperlinks across redirects.  However, we'll need to review the actual `id` tags on the source page to see if the section names match with the original page.

We may also intentionally choose to remove the section tag names because the page ultimately will be quite different from the wiki page.

This Wiki page is intended to be the one-stop shop for those who are
experiencing problems due to this subsystem.

== My Jenkins is broken, help!

If you see an exception like the following in console output, logs,
etc., then you are affected by this access control mechanism:

....
java.lang.SecurityException: agent may not create /var/lib/jenkins/foo/bar/...
See http://jenkins-ci.org/security-144 for more details

java.lang.SecurityException: Sending org.jenkinsci.plugins.gitclient.CliGitAPIImpl$GetPrivateKeys from agent to master is prohibited.
See http://jenkins-ci.org/security-144 for more details
....

The first message indicates that an agent tried to access a file on the
master, and Jenkins has blocked that. It is possible that this indicates
an active attack, but it is also possible that plugins that haven't yet
upgraded to work with SECURITY-144 is doing something legitimate.

The second message indicates that an agent has asked the master to
execute a command, and Jenkins has blocked that because the command
isn't classified by plugin developers whether it is safe to be executed
from the agent.

To resolve this problem, you have the following options:

> * https://wiki.jenkins.io/display/JENKINS//Slave+To+Master+Access+Control#SlaveToMasterAccessControl-update-plugins[Check

Reviewer: This should be just a link to a section, e.g. `<<update-plugins>>`.

Author: @zbynek do you mind breaking down how to link to a section ? I'm thinking on line 76 `<<update-plugins>> [Check ..`

Reviewer: @LarrySul please see https://asciidoctor.org/docs/asciidoc-syntax-quick-reference/#links , the part about *Internal cross references*

if the updates to plugins are available]
* https://wiki.jenkins.io/display/JENKINS//Slave+To+Master+Access+Control#SlaveToMasterAccessControl-kill-switch[Disable
the access control mechanism altogether]
* https://wiki.jenkins.io/display/JENKINS//Slave+To+Master+Access+Control#SlaveToMasterAccessControl-whitelist[Whitelist
the access]

[[SlaveToMasterAccessControl-Checkiftheupdatestopluginsareavailable]]
=== Check if the updates to plugins are available

During the testing of this feature, the following plugins and their
versions have been known to interfere with this subsystem. If you are
using one of those, upgrading those plugins will resolve the problem.
This section will be updated as we discover affected plugins.

[[SlaveToMasterAccessControl-CredentialsPlugin]]

> ==== https://wiki.jenkins.io/display/JENKINS/Credentials+Plugin[Credentials Plugin]

Reviewer: ```suggestion
==== https://plugins.jenkins.io/credentials/[Credentials Plugin]
```

This plugin should be upgraded to 1.17. Earlier versions have command
implementations that are vulnerable to malicious agents.

[[SlaveToMasterAccessControl-SSHCredentialsPlugin]]

> ==== https://wiki.jenkins.io/display/JENKINS/SSH+Credentials+Plugin[SSH Credentials Plugin]

Reviewer: ```suggestion
==== https://plugins.jenkins.io/ssh-credentials/[SSH Credentials Plugin]
```

Reviewer 2: I think it may be better if we drop this section completely.  The plugins referenced are now far beyond those version numbers and Jenkins is now far beyond the base versions described here.

Author: noted

This plugin should be upgraded to 1.10. Earlier versions have command
implementations that are vulnerable to malicious agents.

[[SlaveToMasterAccessControl-GitClientPlugin]]

> ==== https://wiki.jenkins.io/display/JENKINS/Git+Client+Plugin[Git Client Plugin]

Reviewer: ```suggestion
==== link:https://plugins.jenkins.io/git-client/[Git Client Plugin]
```

Author: done

Versions up to 1.10.2 expose vulnerabilities in credentials plugin and
SSH credentials plugin (see above). Upgrading those plugins will fix the
problem, but there's also
https://github.com/jenkinsci/git-client-plugin/pull/147[a separate fix]
in git-client plugin 1.11.0 itself that addresses this problem.

[[SlaveToMasterAccessControl-Enabletheaccesscontrolmechanism]]
=== Enable the access control mechanism

There are essentially three ways of enabling the access control
mechanism:

. Through web UI, go to \http://your-jenkins:8080/configureSecurity and
check "Enable Agent → Master Access Control" option.
. Through file system, create or edit the
file `+$JENKINS_HOME/secrets/slave-to-master-security-kill-switch+` so
that it contains `+false+`.
. Using a

> https://wiki.jenkins-ci.org/display/JENKINS/Groovy+Hook+Script[Groovy Hook Script]

Reviewer: ```suggestion
link:https://www.jenkins.io/doc/book/managing/groovy-hook-scripts/[Groovy Hook Script]
```

> https://wiki.jenkins-ci.org/display/JENKINS/Groovy+Hook+Script[and doing something like this:]

Reviewer: ```suggestion
and doing something like this:
```
Two links to the same page in a single sentence seem redundant.

// true means whitelisted. false or SecurityException means rejected
----

More precisely, `+FilePath+` always internally normalize paths, and
while it allows relative paths, no legitimate code will ever use it, so
it shouldn't have to be factored in when writing rules. It isn't that
the access checking subsystem does normalization/absolutization.

[[SlaveToMasterAccessControl-WhitelistingfromGUI]]
=== Whitelisting from GUI

On Jenkins, you can go to
[.nolink]#http://jenkins/administrativeMonitor/slaveToMasterAccessControl/#
to edit whitelist rules from GUI and have them reflected right away in
the running instance. It consists of the following three sections:

* *Currently whitelisted commands*: See
https://wiki.jenkins.io/display/JENKINS//Slave+To+Master+Access+Control#SlaveToMasterAccessControl-command-whitelist[above]
for what this field means.
* *Currently rejected commands*: This section lists unclassified
commands that Jenkins has actually rejected. You can check boxes and
submit them to have Jenkins write them into the "currently whitelisted
commands" section. Be careful when you do this, though. See

> https://wiki.jenkins.io/display/JENKINS//Slave+To+Master+Access+Control#SlaveToMasterAccessControl-command-whitelist[the

Reviewer: This should be just a link to a section, e.g. `<<command-whitelist>>`.

command whitelisting discussion] above for the implications.
* *File access rules*: See

> https://wiki.jenkins.io/display/JENKINS//Slave+To+Master+Access+Control#SlaveToMasterAccessControl-filepath-rules[above]

Reviewer: This should be just a link to a section, e.g. `<<filepath-rules>>`.

    ...
    public void checkout( ..., FilePath workspace, File _changelogFile ) {
        FilePath changelogFile = new FilePath(_changelogFile);
        workspace.act(new Callable<Void,IOException>() {
            public Void call() {
                // this results in an agent asking the master to open a file for write
                try (OutputStream os = changelogFile.write()) {
                    writeStuffTo(os);
                }
            }
        });
    }
}

// GOOD
class MySCM extends SCM {
    ...
    public void checkout( ..., FilePath workspace, File _changelogFile ) {
        try (final OutputStream out = new RemoteOutputStream(_changelogFile)) {
            workspace.act(new Callable<Void,IOException>() {
                public Void call() {
                    // agent is just writing to a pipe to the master. Quite safe
                    writeStuffTo(out);
                }
            });
        }
    }
}
----

See `+RemoteInputStream+`, `+RemoteOutputStream+`, `+RemoteWriter+`, and
`+Pipe+` for ways to do this.

=== I'm not sure how to adapt my plugin, I need help

If you have questions, please write to jenkinsci-dev@googlegroups.com,

> or talk to us on http://jenkins-ci.org/content/chat[IRC].

Reviewer: ```suggestion
or talk to us on https://www.jenkins.io/chat/[IRC].
```
